### PR TITLE
feat: add Nodeclaim & Machine provision to RAG Engine controller

### PR DESCRIPTION
**Pull request by bangqipropel**
**Reason for Change**:
add Nodeclaim and Machine provision to RAG Engine controller

**Requirements**

- [ ] added unit tests and e2e tests (if applicable).

**Issue Fixed**:

**Notes for Reviewers**:

### CODE REVIEW

File: pkg/controllers/ragengine_status.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	"k8s.io/apimachinery/pkg/api/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (c *RAGEngineReconciler) updateRAGEngineStatus(ctx context.Context, name *client.ObjectKey, condition *metav1.Condition, workerNodes []string) error {
```
Comment (bangqipropel) [inline pkg/controllers/ragengine_status.go:23]: same logic as https://github.com/Azure/kaito/blob/2ec45ac3141e955bcf37af78ba5f76f5f7e65cdd/pkg/controllers/workspace_status.go#L66

File: pkg/controllers/ragengine_controller.go [hunk 2]
```diff
@@ -33,14 +54,381 @@ func NewRAGEngineReconciler(client client.Client, scheme *runtime.Scheme, log lo
...
+func (c *RAGEngineReconciler) addRAGEngine(ctx context.Context, ragEngineObj *kaitov1alpha1.RAGEngine) (reconcile.Result, error) {
+	err := c.applyRAGEngineResource(ctx, ragEngineObj)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
 	return reconcile.Result{}, nil
 }
 
+// applyRAGEngineResource applies RAGEngine resource spec.
+func (c *RAGEngineReconciler) applyRAGEngineResource(ctx context.Context, ragEngineObj *kaitov1alpha1.RAGEngine) error {
```
Comment (bangqipropel) [inline pkg/controllers/ragengine_controller.go:84]: same logic as https://github.com/Azure/kaito/blob/2ec45ac3141e955bcf37af78ba5f76f5f7e65cdd/pkg/controllers/workspace_controller.go#L375

File: pkg/controllers/ragengine_controller.go [hunk 3]
```diff
@@ -33,14 +54,381 @@ func NewRAGEngineReconciler(client client.Client, scheme *runtime.Scheme, log lo
...
+		"ragengineResourceStatusSuccess", "ragengine resource is ready"); err != nil {
+		klog.ErrorS(err, "failed to update ragengine status", "ragengine", klog.KObj(ragEngineObj))
+		return err
+	}
+
+	return nil
+}
+
+// getAllQualifiedNodes returns all nodes that match the labelSelector and instanceType.
+func (c *RAGEngineReconciler) getAllQualifiedNodes(ctx context.Context, ragEngineObj *kaitov1alpha1.RAGEngine) ([]*corev1.Node, error) {
```
Comment (bangqipropel) [inline pkg/controllers/ragengine_controller.go:188]: same logic as https://github.com/Azure/kaito/blob/2ec45ac3141e955bcf37af78ba5f76f5f7e65cdd/pkg/controllers/workspace_controller.go#L478

File: pkg/controllers/ragengine_controller.go [hunk 4]
```diff
@@ -33,14 +54,381 @@ func NewRAGEngineReconciler(client client.Client, scheme *runtime.Scheme, log lo
...
+		if nodeObj.Labels[corev1.LabelInstanceTypeStable] == ragEngineObj.Spec.Compute.InstanceType {
+			qualifiedNodes = append(qualifiedNodes, lo.ToPtr(nodeObj))
+		}
+	}
+
+	return qualifiedNodes, nil
+}
+
+// createAndValidateNode creates a new node and validates status.
+func (c *RAGEngineReconciler) createAndValidateNode(ctx context.Context, ragEngineObj *kaitov1alpha1.RAGEngine) (*corev1.Node, error) {
```
Comment (bangqipropel) [inline pkg/controllers/ragengine_controller.go:234]: similar logic as https://github.com/Azure/kaito/blob/2ec45ac3141e955bcf37af78ba5f76f5f7e65cdd/pkg/controllers/workspace_controller.go#L524

File: pkg/controllers/ragengine_controller.go [hunk 5]
```diff
@@ -33,14 +54,381 @@ func NewRAGEngineReconciler(client client.Client, scheme *runtime.Scheme, log lo
...
+			return nil, updateErr
+		}
+		return nil, err
+	}
+
+	// get the node object from the machine status nodeName.
+	return resources.GetNode(ctx, newMachine.Status.NodeName, c.Client)
+}
+
+func (c *RAGEngineReconciler) CreateNodeClaim(ctx context.Context, ragEngineObj *kaitov1alpha1.RAGEngine, nodeOSDiskSize string) (*corev1.Node, error) {
```
Comment (bangqipropel) [inline pkg/controllers/ragengine_controller.go:283]: same logic to https://github.com/Azure/kaito/blob/2ec45ac3141e955bcf37af78ba5f76f5f7e65cdd/pkg/controllers/workspace_controller.go#L577

File: pkg/controllers/ragengine_controller.go [hunk 6]
```diff
@@ -33,14 +54,381 @@ func NewRAGEngineReconciler(client client.Client, scheme *runtime.Scheme, log lo
...
+	}
+
+	if featuregates.FeatureGates[consts.FeatureFlagKarpenter] {
+		return c.CreateNodeClaim(ctx, ragEngineObj, nodeOSDiskSize)
+	} else {
+		return c.CreateMachine(ctx, ragEngineObj, nodeOSDiskSize)
+	}
+}
+
+func (c *RAGEngineReconciler) CreateMachine(ctx context.Context, ragEngineObj *kaitov1alpha1.RAGEngine, nodeOSDiskSize string) (*corev1.Node, error) {
```
Comment (bangqipropel) [inline pkg/controllers/ragengine_controller.go:248]: same logic to https://github.com/Azure/kaito/blob/2ec45ac3141e955bcf37af78ba5f76f5f7e65cdd/pkg/controllers/workspace_controller.go#L542

File: pkg/controllers/ragengine_controller.go [hunk 7]
```diff
@@ -33,14 +54,381 @@ func NewRAGEngineReconciler(client client.Client, scheme *runtime.Scheme, log lo
...
+		}
+		return nil, err
+	}
+
+	// get the node object from the nodeClaim status nodeName.
+	return resources.GetNode(ctx, newNodeClaim.Status.NodeName, c.Client)
+}
+
+// ensureNodePlugins ensures node plugins are installed.
+func (c *RAGEngineReconciler) ensureNodePlugins(ctx context.Context, ragEngineObj *kaitov1alpha1.RAGEngine, nodeObj *corev1.Node) error {
```
Comment (bangqipropel) [inline pkg/controllers/ragengine_controller.go:319]: same logic to https://github.com/Azure/kaito/blob/2ec45ac3141e955bcf37af78ba5f76f5f7e65cdd/pkg/controllers/workspace_controller.go#L613

File: pkg/controllers/ragengine_controller.go [hunk 8]
```diff
@@ -33,14 +54,381 @@ func NewRAGEngineReconciler(client client.Client, scheme *runtime.Scheme, log lo
...
 		WithOptions(controller.Options{MaxConcurrentReconciles: 5})
+	if featuregates.FeatureGates[consts.FeatureFlagKarpenter] {
+		builder.
+			Watches(&v1beta1.NodeClaim{}, c.watchNodeClaims()) // watches for nodeClaim with labels indicating ragengine name.
+	}
 	return builder.Complete(c)
 }
+
+// watches for machine with labels indicating RAGEngine name.
+func (c *RAGEngineReconciler) watchMachines() handler.EventHandler {
```
Comment (bangqipropel) [inline pkg/controllers/ragengine_controller.go:381]: same logic to https://github.com/Azure/kaito/blob/2ec45ac3141e955bcf37af78ba5f76f5f7e65cdd/pkg/controllers/workspace_controller.go#L890

File: pkg/controllers/ragengine_controller.go [hunk 9]
```diff
@@ -33,14 +54,381 @@ func NewRAGEngineReconciler(client client.Client, scheme *runtime.Scheme, log lo
...
+						Name:      name,
+						Namespace: namespace,
+					},
+				},
+			}
+		})
+}
+
+// watches for nodeClaim with labels indicating RAGEngine name.
+func (c *RAGEngineReconciler) watchNodeClaims() handler.EventHandler {
```
Comment (bangqipropel) [inline pkg/controllers/ragengine_controller.go:413]: same logic to https://github.com/Azure/kaito/blob/2ec45ac3141e955bcf37af78ba5f76f5f7e65cdd/pkg/controllers/workspace_controller.go#L922

**Review by Fei-Guo — APPROVED** [review]

**Pull request merged**